Add game logic for Board cell clicks

davidl21 opened the pull request:
**Summary** 

Implements core game interaction logic upon user clicking a Board cell 

- adds game state validation to prevent out of bounds moves
- sets appropriate cell flags on click
- reveals single or multiple cells according to Minesweeper rules

**Testing**

- adds new unit test
- tests out of bounds edge case
- tests case where a user might click on a revealed cell multiple times
- run tests with `python -m pytest`

**To Do**

- add logic for when player loses the game, revealing all squares
- add error classes for better HTTP responses

davidl21 committed:
added function for handling clicks and reveals

implemented a DFS-based function to reveal neighboring squares and structured function for general clicks

davidl21 committed:
Added basic tests for reveal

Summary: used pytest to add basic tests for reveal on empty and 3x3 boards. Tests that correct cells get revealed, need to add test checking for double counting

davidl21 committed:
Added test to count duplicate reveals

Summary: as per title, added a test to ensure that cells that are already revealed are not revealed again.

davidl21 committed:
Removed unnecessary comments

Summary: as per title

cxindyli commented on backend/models.py:
```diff
@@ -1,4 +1,5 @@
+from typing import Optional
 
 class Cell:
     def __init__(self, x: int, y: int):
```
extract Cell class into different types of cells so that mine and regular-cell functions and attributes can be separated

davidl21 replied:
I'm thinking of leaving the Cell class as one class because it seems like Mine and Non-Mine cells have almost no behavioral differences and it might make typechecking more difficult, and only difference is their behavior when the game ends (when clicking on a mine)

cxindyli commented on backend/models.py:
```diff
@@ -27,6 +28,7 @@ def __init__(self, cols: int, rows: int, num_mines: int):
         self.rows = rows
         self.num_mines = num_mines
         self.grid = [[Cell(r, c) for c in range(self.cols)] for r in range(self.rows)]
+        self.offsets = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, -1), (-1, 1)]
```
don't have to store this on every `Board `instance; instead you can use a class variable

davidl21 replied:
done

cxindyli commented on backend/models.py:
```diff
@@ -27,6 +28,7 @@ def __init__(self, cols: int, rows: int, num_mines: int):
+        self.offsets = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, -1), (-1, 1)]
 
     def place_mines(self):
         all_positions = []
```
for lines 34-40, can extract into another function

cxindyli replied:
also, i don't think it's good to initialize a new instance variable within a random function in a class. you should probably have all instance variables defined clearly in init, so here you could probably just set self.mine_positions to an empty sequence in your init method and then set mine_positions when you get the actual mine positions

davidl21 replied:
since the storing of mine_positions is only useful for the function of placing mines, would it make sense to have it only for the function and not as an instance variable because we don't need it to persist through the game?

maybe for game victory, I'll store the count of revealed cells and the total number of non-mine cells, and check if they're equal to each other?

cxindyli commented on backend/models.py:
```diff
@@ -44,9 +46,7 @@ def place_mines(self):
             self.compute_neighbors(r, c)
 
     def compute_neighbors(self, r: int, c:int) -> None:
```
probably rename this to something else; compute_neighbors gives off the impression that you're simply computing something but you're actually setting some important variables here and making changes to the object

cxindyli replied:
probably rename to `incrementNeighborMineCounts`?

davidl21 replied:
done

cxindyli commented on backend/models.py:
```diff
@@ -44,9 +46,7 @@ def place_mines(self):
             self.compute_neighbors(r, c)
```
i think this should go inside MineCell functionality, where you can just call MineCell.incrementNeighborMineCount

cxindyli replied:
so instead the code would look like:
```suggestion
            for r, c in mine_positions:
              self.grid[r][c] = new MineCell(lalala)
              mine.incrementNeighborMineCount
```

davidl21 replied:
I think a potential issue is that minecells don't have a view of the entire board, so incrementingneighbormine count using r, c indices doesn't work unless I can access the Board's self.grid[][] or something similar

cxindyli commented on backend/models.py:
```diff
@@ -27,6 +28,7 @@ def __init__(self, cols: int, rows: int, num_mines: int):
         self.rows = rows
```
also can't comment on the above code but perhaps rename `add_neighbor `to `increment_neighbor_mine_count`

davidl21 replied:
done

cxindyli commented on backend/models.py:
```diff
@@ -44,9 +46,7 @@ def place_mines(self):
+        for r_offset, c_offset in self.offsets:
             new_r, new_c = r + r_offset, c + c_offset
 
             if self.is_inbound(r, c):
```
don't just return None, should return an Error class that has more specific information

cxindyli commented on backend/models.py:
```diff
@@ -62,30 +62,52 @@ def handle_click(self, r: int, c:int) -> list[Cell]:
-            return
+            return []
+
+    def reveal(self, r: int, c: int) -> list[Cell]:
```
based on this function name it's hard to deduce what it actually does / returns

davidl21 replied:
fixed, changed to get_revealed_cells

cxindyli commented on backend/models.py:
```diff
@@ -62,30 +62,52 @@ def handle_click(self, r: int, c:int) -> list[Cell]:
         revealed_cells = []
 
         if self.is_inbound(r, c):
```
how is it possible for someone to click out of bounds and still give you an (r, c)?

davidl21 replied:
u right...

cxindyli commented on backend/models.py:
```diff
@@ -62,30 +62,52 @@ def handle_click(self, r: int, c:int) -> list[Cell]:
-                return
+                cell.toggle_flag()
 
             if cell.is_mine:
```
here you'd check for the type of cell instead of just boolean like so:

```
cell match {
case MineCell => Result(Result.GameOver)
case RegularCell => self.reveal 
                                 Result(Result.Continue)
}
```
store list[revealed] as an attribute for the board, and then change as it's changed

davidl21 committed:
Made offsets array into class variable

davidl21 committed:
Created new helper function to return all coordinates in board

Summary: moved functionality of finding all x,y coordinates to a helper functino to be used later in random sampling of mine placing

davidl21 committed:
Renamed compute_neighbors to increment_neighbor_mine_counts

davidl21 committed:
Renamed add_neighbor to increment_neighbor_mine_count

davidl21 committed:
Renamed vague reveal function to get_revealed_cells

davidl21 committed:
Removed unnecessary out of bounds check for handle_click

Summary: when a user clicks out of bounds we won't be providing handle click with an r, c input anyways, so the check was redundant and is removed
